File: Advent2022/Day2/Day2.py
import aoc_utils as au


def score(them: str, me: str):
    return [[4, 8, 3], [1, 5, 9], [7, 2, 6]][ord(them) - ord("A")][ord(me) - ord("X")]


# expected 15
def test_one(input: list[str]) -> int:
    return sum([score(*game.split()) for game in input])


# expected 10404
def part_one(input: list[str]) -> int:
    return sum([score(*game.split()) for game in input])


def score2(them: str, me: str):
    return [[3, 4, 8], [1, 5, 9], [2, 6, 7]][ord(them) - ord("A")][ord(me) - ord("X")]


# expected 12
def test_two(input: list[str]) -> int:
    return sum([score2(*game.split()) for game in input])


# expected 10334
def part_two(input: list[str]) -> int:
    return sum([score2(*game.split()) for game in input])


if __name__ == "__main__":
    day = au.Day(2022, 2, test_one, test_two, part_one, part_two)

    day.run_all(run_tests=True)

File: Advent2022/Day2/Day2.cpp
#include "aoc_utils.h"

#include <string>
#include <vector>
#include <algorithm>
#include <numeric>

#include <iostream>

using namespace std;

class Day2 : public au::Day<vector<string>>
{
private:
    int score(const string &game)
    {
        int scoremap[3][3] = {
            {4, 8, 3},
            {1, 5, 9},
            {7, 2, 6}};

        auto them = game.c_str()[0];
        auto me = game.c_str()[2];
        return scoremap[them - 'A'][me - 'X'];
    }

    int score2(const string &game)
    {
        int scoremap[3][3] = {
            {3, 4, 8},
            {1, 5, 9},
            {2, 6, 7}};

        auto them = game.c_str()[0];
        auto me = game.c_str()[2];
        return scoremap[them - 'A'][me - 'X'];
    }

public:
    Day2() : Day(2022, 2) {}

    int test_one(const vector<string> &input)
    {
        auto sm = 0;
        for (auto game : input)
            sm += score(game);
        return sm;
    }

    int part_one(const vector<string> &input)
    {
        int sm = 0;
        for (auto game : input)
            sm += score(game);
        return sm;
    }

    int test_two(const vector<string> &input)
    {
        auto sm = 0;
        for (auto game : input)
            sm += score2(game);
        return sm;
    }

    int part_two(const vector<string> &input)
    {
        auto sm = 0;
        for (auto game : input)
            sm += score2(game);
        return sm;
    }
};

int main()
{
    Day2().run_all(true);
}
